### DISCUSSION TITLE
Composite-Device: Multiple different reports without delay

### DISCUSSION THREAD

**Discussion by liam-lemonade** [adafruit/Adafruit_TinyUSB_Arduino#282] [Q&A]
How can I send reports of different types and IDs without delaying inbetween? [hid_composite.ino](https://github.com/adafruit/Adafruit_TinyUSB_Arduino/blob/master/examples/HID/hid_composite/hid_composite.ino) uses `delay(10)` to achieve results, but this sacrifices the report rate of my device. How can I do this (in any manner, even completely different approaches) without using any delay?

My current sketch:
```cpp
#include "Adafruit_TinyUSB.h"
#include <Arduino.h>

constexpr int RID_KEYBOARD = 1, RID_MOUSE = 2;

constexpr uint8_t HID_REPORT_DESC[] = {
    TUD_HID_REPORT_DESC_KEYBOARD(HID_REPORT_ID(RID_KEYBOARD)),
    TUD_HID_REPORT_DESC_MOUSE(HID_REPORT_ID(RID_MOUSE)),
};

// host polls the buffer @ 1000hz
Adafruit_USBD_HID usb_hid(HID_REPORT_DESC, sizeof(HID_REPORT_DESC), HID_ITF_PROTOCOL_NONE, 1, false);

void setup() {
    usb_hid.setStringDescriptor("Testing HID Device");
    usb_hid.begin();

    Serial.begin(115200);

    pinMode(LED_BUILTIN, OUTPUT);

    while(!TinyUSBDevice.mounted())
        delay(1);
}

void loop() {
    delay(1000);

    uint8_t keycode[6] = { 0 };
    keycode[0] = HID_KEY_A;

    usb_hid.keyboardReport(RID_KEYBOARD, 0, keycode); // works
    usb_hid.mouseMove(RID_MOUSE, 100, 100); // skipped
    digitalWrite(LED_BUILTIN, HIGH);

    delay(1000);

    keycode[0] = 0;
    usb_hid.keyboardReport(RID_KEYBOARD, 0, keycode); // works
    usb_hid.mouseMove(RID_MOUSE, -100, -100); // also skipped
    digitalWrite(LED_BUILTIN, LOW);
}
```

**Comment by hathach** [comment]
each HID interface can typically only send 1 report per polling interval e.g 2 or 8 ms. Therefore, even if you stack them together, it won't send that fast. Your other option is using multiple interfaces instead of composite one. I just enable dual HID interfaces #280, but haven't got time to add example sketch, which mcu you are using ? I will try to add one whenever I could.

**Comment by liam-lemonade** [comment]
I am currently running on an RP2040, in particular a Raspberry Pi Pico. I am running the [earlephilhower/arduino-pico](https://github.com/earlephilhower/arduino-pico) core.

**Comment by hathach** [comment]
should be possible with https://github.com/adafruit/Adafruit_TinyUSB_Arduino/releases/tag/2.1.0 (2 interfaces). Also an example hid_dual_interfaces is also added.
